macros/anaUltraThreshold.py

```diff
@@ -53,121 +64,215 @@
-        if event.vth1 > 150 and event.Nhits > 0:
-            hot_channels[event.vfatN][pan_pin] = True
-            pass
         pass
     else:
         vSum[event.vfatN].Fill(strip,event.vth1,event.Nhits)
-        if event.vth1 > 150 and event.Nhits > 0:
-            hot_channels[event.vfatN][strip] = True
-            pass
         pass
 
+#Determine Hot Channels
+print 'Determining hot channels'
+from qcutilities import isOutlier
+import numpy as np
+for vfat in range(0,24):
+    if (vfat_mask >> vfat) & 0x1: continue
+
+    #For each channel determine the maximum thresholds
+    chanMaxVT1 = np.zeros((2,vSum[vfat].GetNbinsX()))
+    for chan in range(0,vSum[vfat].GetNbinsX()):
+	      for thresh in range(VT1_MAX+1,0,-1):
+	          if(vSum[vfat].ProjectionY("projY",chan,chan,"").GetBinContent(thresh+1) > 1.0):
+		            #print chan, j
+		            chanMaxVT1[0][chan]=chan
+		            chanMaxVT1[1][chan]=(thresh+1)
+		            break
+	          pass
+	      pass
+
+    #Determine Outliers (e.g. "hot" channels)
+    chanOutliers = isOutlier(chanMaxVT1[1,:])
+    for chan in range(0,len(chanOutliers)):
+	      hot_channels[vfat][chan] = chanOutliers[chan]
+	      pass
+
+    if options.debug:
+	      print "VFAT%i Max Thresholds By Channel"%vfat
+	      print chanMaxVT1
+
+	      print "VFAT%i Channel Outliers"%vfat
+	      chanOutliers = np.column_stack((chanMaxVT1[0,:],np.array(hot_channels[vfat]).astype(float)))
+	      print chanOutliers
+
+	      pass
+    pass
+
+# Fetch trimDAC & chMask from scurveFitTree
+import root_numpy as rp #note need root_numpy-4.7.2 (may need to run 'pip install root_numpy --upgrade')
+from qcutilities import initVFATArray
+dict_vfatTrimMaskData = {}
+if options.chConfigKnown:
+    list_bNames = ["vfatN"]
+    if not (options.channels or options.PanPin):
+        list_bNames.append("ROBstr")
+    	  pass
+    elif options.channels:
+    	  list_bNames.append("vfatCh")
+    	  pass
+    elif options.PanPin:
+    	  list_bNames.append("panPin")
+    	  pass
+    list_bNames.append("mask")
+    list_bNames.append("trimDAC")
+
+    try:
+    	  array_VFATSCurveData = rp.root2array(options.fileScurveFitTree,treename="scurveFitTree",branches=list_bNames)
+
+    	  #Store array_VFATSCurveData into a dict for easy access
+    	  #This dictionary has VFAT position as the key value, returns a structured numpy array
+    	  dict_vfatTrimMaskData = {idx:initVFATArray(array_VFATSCurveData.dtype) for idx in np.unique(array_VFATSCurveData[list_bNames[0]])}
+    	  for dataPt in array_VFATSCurveData:
+	          dict_vfatTrimMaskData[dataPt['vfatN']][dataPt[list_bNames[1]]]['mask'] =  dataPt['mask']
+            dict_vfatTrimMaskData[dataPt['vfatN']][dataPt[list_bNames[1]]]['trimDAC'] =  dataPt['trimDAC']
+	          pass
+    	  pass
+    except Exception as e:
+    	  print '%s does not seem to exist'%options.fileScurveFitTree
+    	  print e
+    	  pass
+    pass
+
+#Save Output
```

Comment:
lots of strange indenting, try to configure editor to automatically convert tabs to spaces (looks like 4 is the default for these python files)